nemo/export/trt_llm/decoder/__init__.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""LLM Decoder implementation for tensorrt_llm conversion."""
+from typing import Dict, Type
+
+import tensorrt as trt
+
+from ..model_config import (
+    DECODER_FALCON,
+    DECODER_GEMMA,
+    DECODER_GPT2,
+    DECODER_GPTJ,
+    DECODER_GPTNEXT,
+    DECODER_LLAMA,
+    QUANTIZATION_NONE,
+)
+from .decoder import DecoderLayerBuilder, DecoderLayerConfigBuilder
+from .falcon import FALCONDecoderLayerBuilder, FALCONDecoderLayerConfigBuilder
+from .gemma import GemmaDecoderLayerBuilder, GemmaDecoderLayerConfigBuilder
+from .gpt import GPTDecoderLayerBuilder, GPTDecoderLayerConfigBuilder
+from .gptj import GPTJDecoderLayerBuilder, GPTJDecoderLayerConfigBuilder
+from .llama import LLAMADecoderLayerBuilder, LLAMADecoderLayerConfigBuilder
+
+DECODER_CONFIG_REGISTRY: Dict[str, Type[DecoderLayerConfigBuilder]] = {
+    DECODER_GPT2: GPTDecoderLayerConfigBuilder,
+    DECODER_GPTJ: GPTJDecoderLayerConfigBuilder,
+    DECODER_LLAMA: LLAMADecoderLayerConfigBuilder,
+    DECODER_FALCON: FALCONDecoderLayerConfigBuilder,
+    DECODER_GEMMA: GemmaDecoderLayerConfigBuilder,
+}
+
+
+def build_decoder_layer_config(layer, decoder: str, dtype=trt.float16, rank=0, tensor_parallel=1):
+    """Builds the decoder layer config with the input torch module."""
+    assert decoder in DECODER_CONFIG_REGISTRY, f"{decoder} not supported"
+    return DECODER_CONFIG_REGISTRY[decoder](decoder, dtype, rank, tensor_parallel).build_layer(layer)
+
+
+DECODER_REGISTRY: Dict[str, Type[DecoderLayerBuilder]] = {
```

Comment:
I like this, but can we have a static method that updates the registry dynamically rather than hard coding it by hand here? For example, have a method that takes in a class, uses class name to register inside this dict if not exists, else raise error if already registered.

We follow this kind of dynamic registry updating inside of NeMo's optimizer / scheduler codebase, plus a few other domains.

Reply:
This part comes from the TRT-LLM team. I really rather not to touch here at all.

Reply:
Hmm ok